otter/convergence/steps.py

```diff
@@ -147,17 +98,24 @@ def report_success(result):
 
         def report_failure(result):
             """
-            If the failure is an APIError with a recognized user error,
-            return a :obj:`StepResult.FAILURE` along with that user error as
-            a message, otherwise return a :obj:`StepResult.RETRY` without
-            a message.
+            If the failure is a :class:`CreateServerConfigurationError` or a
+            :class:`CreateServerOverQuoteError`, or a :class:`APIError` with
+            a 4XX status code, return a :obj:`StepResult.FAILURE` along with
+            that error a areason.
+
+            Otherwise, all other errors are considered non-terminal and a
+            :obj:`StepResult.RETRY` is returned.
             """
             err_type, error, traceback = result
-            if err_type == APIError:
-                message = _parse_nova_user_error(error)
-                if message is not None:
-                    return StepResult.FAILURE, [ErrorReason.String(message)]
 
+            terminal_error = (
```

Comment:
Note to reviewer: I plan on refactoring this (and in the `AddNodesToCLB` step) in the next PR.